{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ee9b0645",
   "metadata": {},
   "source": [
    "# Final assignment Unsupervised Learning"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8db88e8",
   "metadata": {},
   "source": [
    "## Background information\n",
    "\n",
    "This data is from a study of [NSCLC-Radiomics-Genomics](https://wiki.cancerimagingarchive.net/display/Public/NSCLC-Radiomics-Genomics#16056856db10d39adf704eefa53e41edcf5ef41c) and it contains information on 89 patients with non-small cell lung cancer (NSCLC) that were treated with surgery. There are two files, one containing gene expression data and another containing metadata. Both need to be fused before beginning analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9a502d7",
   "metadata": {},
   "source": [
    "# Loading in the data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763ef66e",
   "metadata": {},
   "source": [
    "### Gene Expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "00a7dc5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pylab\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aaf6e7b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID_REF</th>\n",
       "      <th>AFFX-BioB-3_at</th>\n",
       "      <th>AFFX-BioB-5_at</th>\n",
       "      <th>AFFX-BioB-M_at</th>\n",
       "      <th>AFFX-BioC-3_at</th>\n",
       "      <th>AFFX-BioC-5_at</th>\n",
       "      <th>AFFX-BioDn-3_at</th>\n",
       "      <th>AFFX-BioDn-5_at</th>\n",
       "      <th>AFFX-BkGr-GC03_at</th>\n",
       "      <th>AFFX-BkGr-GC04_at</th>\n",
       "      <th>AFFX-BkGr-GC05_at</th>\n",
       "      <th>...</th>\n",
       "      <th>merck2-Z41436_at</th>\n",
       "      <th>merck2-Z43503_at</th>\n",
       "      <th>merck2-Z47244_x_at</th>\n",
       "      <th>merck2-Z47250_x_at</th>\n",
       "      <th>merck2-Z48501_s_at</th>\n",
       "      <th>merck2-Z48633_s_at</th>\n",
       "      <th>merck2-Z70222_a_at</th>\n",
       "      <th>merck2-Z70608_x_at</th>\n",
       "      <th>merck2-Z72499_a_at</th>\n",
       "      <th>merck2-Z75331_a_at</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1416528</th>\n",
       "      <td>7.376915</td>\n",
       "      <td>6.984530</td>\n",
       "      <td>7.330576</td>\n",
       "      <td>10.922741</td>\n",
       "      <td>11.032030</td>\n",
       "      <td>13.411168</td>\n",
       "      <td>12.625445</td>\n",
       "      <td>1.464847</td>\n",
       "      <td>1.416673</td>\n",
       "      <td>1.438285</td>\n",
       "      <td>...</td>\n",
       "      <td>7.077955</td>\n",
       "      <td>4.350631</td>\n",
       "      <td>7.506299</td>\n",
       "      <td>9.774634</td>\n",
       "      <td>11.295187</td>\n",
       "      <td>10.286077</td>\n",
       "      <td>6.892053</td>\n",
       "      <td>1.777216</td>\n",
       "      <td>9.031084</td>\n",
       "      <td>7.284069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416529</th>\n",
       "      <td>8.024915</td>\n",
       "      <td>7.427048</td>\n",
       "      <td>8.010530</td>\n",
       "      <td>11.390638</td>\n",
       "      <td>11.533338</td>\n",
       "      <td>13.546671</td>\n",
       "      <td>12.898906</td>\n",
       "      <td>1.502126</td>\n",
       "      <td>1.398866</td>\n",
       "      <td>1.403564</td>\n",
       "      <td>...</td>\n",
       "      <td>7.552333</td>\n",
       "      <td>3.578359</td>\n",
       "      <td>5.516440</td>\n",
       "      <td>4.601167</td>\n",
       "      <td>11.443442</td>\n",
       "      <td>11.173715</td>\n",
       "      <td>5.513603</td>\n",
       "      <td>1.940099</td>\n",
       "      <td>9.527973</td>\n",
       "      <td>6.985508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416530</th>\n",
       "      <td>7.522543</td>\n",
       "      <td>7.077207</td>\n",
       "      <td>7.334551</td>\n",
       "      <td>10.936703</td>\n",
       "      <td>11.018510</td>\n",
       "      <td>13.360017</td>\n",
       "      <td>12.540647</td>\n",
       "      <td>1.355337</td>\n",
       "      <td>1.343599</td>\n",
       "      <td>1.306212</td>\n",
       "      <td>...</td>\n",
       "      <td>5.084856</td>\n",
       "      <td>2.937384</td>\n",
       "      <td>6.310065</td>\n",
       "      <td>6.315107</td>\n",
       "      <td>11.584115</td>\n",
       "      <td>10.586540</td>\n",
       "      <td>6.112918</td>\n",
       "      <td>1.753519</td>\n",
       "      <td>9.515653</td>\n",
       "      <td>6.556233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416531</th>\n",
       "      <td>7.152864</td>\n",
       "      <td>6.849513</td>\n",
       "      <td>7.143286</td>\n",
       "      <td>10.791909</td>\n",
       "      <td>10.875259</td>\n",
       "      <td>13.364171</td>\n",
       "      <td>12.579293</td>\n",
       "      <td>1.327548</td>\n",
       "      <td>1.272961</td>\n",
       "      <td>1.354529</td>\n",
       "      <td>...</td>\n",
       "      <td>7.136409</td>\n",
       "      <td>2.904756</td>\n",
       "      <td>5.954062</td>\n",
       "      <td>6.738276</td>\n",
       "      <td>11.240300</td>\n",
       "      <td>10.257621</td>\n",
       "      <td>6.276813</td>\n",
       "      <td>1.889742</td>\n",
       "      <td>9.383670</td>\n",
       "      <td>7.293852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416532</th>\n",
       "      <td>7.211031</td>\n",
       "      <td>6.753131</td>\n",
       "      <td>7.077163</td>\n",
       "      <td>10.701328</td>\n",
       "      <td>10.823792</td>\n",
       "      <td>13.340075</td>\n",
       "      <td>12.420624</td>\n",
       "      <td>1.296788</td>\n",
       "      <td>1.244117</td>\n",
       "      <td>1.291959</td>\n",
       "      <td>...</td>\n",
       "      <td>7.667150</td>\n",
       "      <td>3.119091</td>\n",
       "      <td>7.257312</td>\n",
       "      <td>8.118139</td>\n",
       "      <td>10.998377</td>\n",
       "      <td>10.634072</td>\n",
       "      <td>7.041829</td>\n",
       "      <td>1.686633</td>\n",
       "      <td>9.464486</td>\n",
       "      <td>6.244336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416612</th>\n",
       "      <td>7.215856</td>\n",
       "      <td>6.740992</td>\n",
       "      <td>7.152722</td>\n",
       "      <td>10.707980</td>\n",
       "      <td>10.842697</td>\n",
       "      <td>13.332414</td>\n",
       "      <td>12.397369</td>\n",
       "      <td>1.443620</td>\n",
       "      <td>1.342957</td>\n",
       "      <td>1.380656</td>\n",
       "      <td>...</td>\n",
       "      <td>6.044094</td>\n",
       "      <td>3.333540</td>\n",
       "      <td>8.953402</td>\n",
       "      <td>10.221695</td>\n",
       "      <td>11.252406</td>\n",
       "      <td>11.010846</td>\n",
       "      <td>6.493021</td>\n",
       "      <td>1.924313</td>\n",
       "      <td>9.223950</td>\n",
       "      <td>5.817185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416613</th>\n",
       "      <td>7.465109</td>\n",
       "      <td>7.080787</td>\n",
       "      <td>7.422373</td>\n",
       "      <td>10.736531</td>\n",
       "      <td>10.900509</td>\n",
       "      <td>13.285086</td>\n",
       "      <td>12.356609</td>\n",
       "      <td>1.491890</td>\n",
       "      <td>1.450109</td>\n",
       "      <td>1.436852</td>\n",
       "      <td>...</td>\n",
       "      <td>5.277468</td>\n",
       "      <td>2.861520</td>\n",
       "      <td>6.342812</td>\n",
       "      <td>6.896812</td>\n",
       "      <td>11.694623</td>\n",
       "      <td>9.975013</td>\n",
       "      <td>7.952822</td>\n",
       "      <td>1.783032</td>\n",
       "      <td>9.481911</td>\n",
       "      <td>7.487134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416614</th>\n",
       "      <td>7.245458</td>\n",
       "      <td>6.725812</td>\n",
       "      <td>7.263596</td>\n",
       "      <td>10.770440</td>\n",
       "      <td>10.838008</td>\n",
       "      <td>13.469373</td>\n",
       "      <td>12.566290</td>\n",
       "      <td>1.441152</td>\n",
       "      <td>1.348752</td>\n",
       "      <td>1.356916</td>\n",
       "      <td>...</td>\n",
       "      <td>7.196013</td>\n",
       "      <td>3.903299</td>\n",
       "      <td>5.724736</td>\n",
       "      <td>5.820700</td>\n",
       "      <td>11.692384</td>\n",
       "      <td>10.044351</td>\n",
       "      <td>6.978308</td>\n",
       "      <td>1.371800</td>\n",
       "      <td>9.632680</td>\n",
       "      <td>7.410402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416615</th>\n",
       "      <td>7.039592</td>\n",
       "      <td>6.575376</td>\n",
       "      <td>6.975162</td>\n",
       "      <td>10.367009</td>\n",
       "      <td>10.494806</td>\n",
       "      <td>13.214522</td>\n",
       "      <td>12.108750</td>\n",
       "      <td>1.460382</td>\n",
       "      <td>1.362631</td>\n",
       "      <td>1.360318</td>\n",
       "      <td>...</td>\n",
       "      <td>2.952597</td>\n",
       "      <td>2.844055</td>\n",
       "      <td>7.236876</td>\n",
       "      <td>8.625332</td>\n",
       "      <td>10.826930</td>\n",
       "      <td>8.798358</td>\n",
       "      <td>7.460604</td>\n",
       "      <td>1.811532</td>\n",
       "      <td>9.313734</td>\n",
       "      <td>6.396247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1416616</th>\n",
       "      <td>7.393667</td>\n",
       "      <td>7.041106</td>\n",
       "      <td>7.366620</td>\n",
       "      <td>10.753221</td>\n",
       "      <td>10.863600</td>\n",
       "      <td>13.243353</td>\n",
       "      <td>12.242187</td>\n",
       "      <td>1.564256</td>\n",
       "      <td>1.342802</td>\n",
       "      <td>1.310207</td>\n",
       "      <td>...</td>\n",
       "      <td>5.835789</td>\n",
       "      <td>3.496999</td>\n",
       "      <td>8.771954</td>\n",
       "      <td>10.271560</td>\n",
       "      <td>10.663018</td>\n",
       "      <td>10.444210</td>\n",
       "      <td>7.519374</td>\n",
       "      <td>2.444065</td>\n",
       "      <td>9.527973</td>\n",
       "      <td>5.605377</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 60607 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID_REF      AFFX-BioB-3_at  AFFX-BioB-5_at  AFFX-BioB-M_at  AFFX-BioC-3_at  \\\n",
       "GSM1416528        7.376915        6.984530        7.330576       10.922741   \n",
       "GSM1416529        8.024915        7.427048        8.010530       11.390638   \n",
       "GSM1416530        7.522543        7.077207        7.334551       10.936703   \n",
       "GSM1416531        7.152864        6.849513        7.143286       10.791909   \n",
       "GSM1416532        7.211031        6.753131        7.077163       10.701328   \n",
       "...                    ...             ...             ...             ...   \n",
       "GSM1416612        7.215856        6.740992        7.152722       10.707980   \n",
       "GSM1416613        7.465109        7.080787        7.422373       10.736531   \n",
       "GSM1416614        7.245458        6.725812        7.263596       10.770440   \n",
       "GSM1416615        7.039592        6.575376        6.975162       10.367009   \n",
       "GSM1416616        7.393667        7.041106        7.366620       10.753221   \n",
       "\n",
       "ID_REF      AFFX-BioC-5_at  AFFX-BioDn-3_at  AFFX-BioDn-5_at  \\\n",
       "GSM1416528       11.032030        13.411168        12.625445   \n",
       "GSM1416529       11.533338        13.546671        12.898906   \n",
       "GSM1416530       11.018510        13.360017        12.540647   \n",
       "GSM1416531       10.875259        13.364171        12.579293   \n",
       "GSM1416532       10.823792        13.340075        12.420624   \n",
       "...                    ...              ...              ...   \n",
       "GSM1416612       10.842697        13.332414        12.397369   \n",
       "GSM1416613       10.900509        13.285086        12.356609   \n",
       "GSM1416614       10.838008        13.469373        12.566290   \n",
       "GSM1416615       10.494806        13.214522        12.108750   \n",
       "GSM1416616       10.863600        13.243353        12.242187   \n",
       "\n",
       "ID_REF      AFFX-BkGr-GC03_at  AFFX-BkGr-GC04_at  AFFX-BkGr-GC05_at  ...  \\\n",
       "GSM1416528           1.464847           1.416673           1.438285  ...   \n",
       "GSM1416529           1.502126           1.398866           1.403564  ...   \n",
       "GSM1416530           1.355337           1.343599           1.306212  ...   \n",
       "GSM1416531           1.327548           1.272961           1.354529  ...   \n",
       "GSM1416532           1.296788           1.244117           1.291959  ...   \n",
       "...                       ...                ...                ...  ...   \n",
       "GSM1416612           1.443620           1.342957           1.380656  ...   \n",
       "GSM1416613           1.491890           1.450109           1.436852  ...   \n",
       "GSM1416614           1.441152           1.348752           1.356916  ...   \n",
       "GSM1416615           1.460382           1.362631           1.360318  ...   \n",
       "GSM1416616           1.564256           1.342802           1.310207  ...   \n",
       "\n",
       "ID_REF      merck2-Z41436_at  merck2-Z43503_at  merck2-Z47244_x_at  \\\n",
       "GSM1416528          7.077955          4.350631            7.506299   \n",
       "GSM1416529          7.552333          3.578359            5.516440   \n",
       "GSM1416530          5.084856          2.937384            6.310065   \n",
       "GSM1416531          7.136409          2.904756            5.954062   \n",
       "GSM1416532          7.667150          3.119091            7.257312   \n",
       "...                      ...               ...                 ...   \n",
       "GSM1416612          6.044094          3.333540            8.953402   \n",
       "GSM1416613          5.277468          2.861520            6.342812   \n",
       "GSM1416614          7.196013          3.903299            5.724736   \n",
       "GSM1416615          2.952597          2.844055            7.236876   \n",
       "GSM1416616          5.835789          3.496999            8.771954   \n",
       "\n",
       "ID_REF      merck2-Z47250_x_at  merck2-Z48501_s_at  merck2-Z48633_s_at  \\\n",
       "GSM1416528            9.774634           11.295187           10.286077   \n",
       "GSM1416529            4.601167           11.443442           11.173715   \n",
       "GSM1416530            6.315107           11.584115           10.586540   \n",
       "GSM1416531            6.738276           11.240300           10.257621   \n",
       "GSM1416532            8.118139           10.998377           10.634072   \n",
       "...                        ...                 ...                 ...   \n",
       "GSM1416612           10.221695           11.252406           11.010846   \n",
       "GSM1416613            6.896812           11.694623            9.975013   \n",
       "GSM1416614            5.820700           11.692384           10.044351   \n",
       "GSM1416615            8.625332           10.826930            8.798358   \n",
       "GSM1416616           10.271560           10.663018           10.444210   \n",
       "\n",
       "ID_REF      merck2-Z70222_a_at  merck2-Z70608_x_at  merck2-Z72499_a_at  \\\n",
       "GSM1416528            6.892053            1.777216            9.031084   \n",
       "GSM1416529            5.513603            1.940099            9.527973   \n",
       "GSM1416530            6.112918            1.753519            9.515653   \n",
       "GSM1416531            6.276813            1.889742            9.383670   \n",
       "GSM1416532            7.041829            1.686633            9.464486   \n",
       "...                        ...                 ...                 ...   \n",
       "GSM1416612            6.493021            1.924313            9.223950   \n",
       "GSM1416613            7.952822            1.783032            9.481911   \n",
       "GSM1416614            6.978308            1.371800            9.632680   \n",
       "GSM1416615            7.460604            1.811532            9.313734   \n",
       "GSM1416616            7.519374            2.444065            9.527973   \n",
       "\n",
       "ID_REF      merck2-Z75331_a_at  \n",
       "GSM1416528            7.284069  \n",
       "GSM1416529            6.985508  \n",
       "GSM1416530            6.556233  \n",
       "GSM1416531            7.293852  \n",
       "GSM1416532            6.244336  \n",
       "...                        ...  \n",
       "GSM1416612            5.817185  \n",
       "GSM1416613            7.487134  \n",
       "GSM1416614            7.410402  \n",
       "GSM1416615            6.396247  \n",
       "GSM1416616            5.605377  \n",
       "\n",
       "[89 rows x 60607 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('data/GSE58661_series_matrix.txt.gz', compression='gzip', header=61, sep='\\t', quotechar='\"')\n",
    "df = df.set_index('ID_REF').transpose().drop('!series_matrix_table_end', axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ac52d30b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There seem to be 0.0 NaNs in the dataframe\n"
     ]
    }
   ],
   "source": [
    "print('There seem to be {:.1f} NaNs in the dataframe'.format(df.isna().sum().sum()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1288f94",
   "metadata": {},
   "source": [
    "### Meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5def0778",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample.name</th>\n",
       "      <th>title</th>\n",
       "      <th>CEL.file</th>\n",
       "      <th>source.location</th>\n",
       "      <th>organism</th>\n",
       "      <th>characteristics.tag.gender</th>\n",
       "      <th>characteristics.tag.histology</th>\n",
       "      <th>characteristics.tag.tumor.size.maximumdiameter</th>\n",
       "      <th>characteristics.tag.stage.primary.tumor</th>\n",
       "      <th>characteristics.tag.stage.nodes</th>\n",
       "      <th>characteristics.tag.stage.mets</th>\n",
       "      <th>characteristics.tag.primaryVSmets</th>\n",
       "      <th>characteristics.tag.grade</th>\n",
       "      <th>molecule tested</th>\n",
       "      <th>label</th>\n",
       "      <th>platform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LUNG3-01</td>\n",
       "      <td>lung_1</td>\n",
       "      <td>LUNG3-01.CEL</td>\n",
       "      <td>Left  Lower Lobe</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>M</td>\n",
       "      <td>Squamous Cell Carcinoma, NOS</td>\n",
       "      <td>4.0</td>\n",
       "      <td>pT2</td>\n",
       "      <td>pN0</td>\n",
       "      <td>pM0</td>\n",
       "      <td>Primary</td>\n",
       "      <td>3</td>\n",
       "      <td>total RNA</td>\n",
       "      <td>biotin</td>\n",
       "      <td>GPL15048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LUNG3-02</td>\n",
       "      <td>lung_2</td>\n",
       "      <td>LUNG3-02.CEL</td>\n",
       "      <td>Left  Lower Lobe</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>M</td>\n",
       "      <td>Adenocarcinoma, Papillary, NOS</td>\n",
       "      <td>1.3</td>\n",
       "      <td>pT1</td>\n",
       "      <td>pNX</td>\n",
       "      <td>pMX</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>total RNA</td>\n",
       "      <td>biotin</td>\n",
       "      <td>GPL15048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LUNG3-03</td>\n",
       "      <td>lung_3</td>\n",
       "      <td>LUNG3-03.CEL</td>\n",
       "      <td>Left  Lower Lobe</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>M</td>\n",
       "      <td>Non-Small Cell</td>\n",
       "      <td>11.0</td>\n",
       "      <td>pT3</td>\n",
       "      <td>pN0</td>\n",
       "      <td>pM0</td>\n",
       "      <td>Primary</td>\n",
       "      <td>3</td>\n",
       "      <td>total RNA</td>\n",
       "      <td>biotin</td>\n",
       "      <td>GPL15048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LUNG3-04</td>\n",
       "      <td>lung_4</td>\n",
       "      <td>LUNG3-04.CEL</td>\n",
       "      <td>Left  Lower Lobe</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>M</td>\n",
       "      <td>Papillary Type AND      Adenocarcinoma, Bronch...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pTX</td>\n",
       "      <td>pNx</td>\n",
       "      <td>pM1</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>total RNA</td>\n",
       "      <td>biotin</td>\n",
       "      <td>GPL15048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LUNG3-05</td>\n",
       "      <td>lung_5</td>\n",
       "      <td>LUNG3-05.CEL</td>\n",
       "      <td>Left  Lower Lobe</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>F</td>\n",
       "      <td>Squamous Cell Carcinoma, NOS</td>\n",
       "      <td>7.8</td>\n",
       "      <td>pT3</td>\n",
       "      <td>pN0</td>\n",
       "      <td>pM0</td>\n",
       "      <td>Primary</td>\n",
       "      <td>2</td>\n",
       "      <td>total RNA</td>\n",
       "      <td>biotin</td>\n",
       "      <td>GPL15048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample.name   title      CEL.file   source.location      organism  \\\n",
       "0    LUNG3-01  lung_1  LUNG3-01.CEL  Left  Lower Lobe  Homo sapiens   \n",
       "1    LUNG3-02  lung_2  LUNG3-02.CEL  Left  Lower Lobe  Homo sapiens   \n",
       "2    LUNG3-03  lung_3  LUNG3-03.CEL  Left  Lower Lobe  Homo sapiens   \n",
       "3    LUNG3-04  lung_4  LUNG3-04.CEL  Left  Lower Lobe  Homo sapiens   \n",
       "4    LUNG3-05  lung_5  LUNG3-05.CEL  Left  Lower Lobe  Homo sapiens   \n",
       "\n",
       "  characteristics.tag.gender  \\\n",
       "0                          M   \n",
       "1                          M   \n",
       "2                          M   \n",
       "3                          M   \n",
       "4                          F   \n",
       "\n",
       "                       characteristics.tag.histology  \\\n",
       "0                       Squamous Cell Carcinoma, NOS   \n",
       "1                     Adenocarcinoma, Papillary, NOS   \n",
       "2                                     Non-Small Cell   \n",
       "3  Papillary Type AND      Adenocarcinoma, Bronch...   \n",
       "4                       Squamous Cell Carcinoma, NOS   \n",
       "\n",
       "   characteristics.tag.tumor.size.maximumdiameter  \\\n",
       "0                                             4.0   \n",
       "1                                             1.3   \n",
       "2                                            11.0   \n",
       "3                                             NaN   \n",
       "4                                             7.8   \n",
       "\n",
       "  characteristics.tag.stage.primary.tumor characteristics.tag.stage.nodes  \\\n",
       "0                                     pT2                             pN0   \n",
       "1                                     pT1                             pNX   \n",
       "2                                     pT3                             pN0   \n",
       "3                                     pTX                             pNx   \n",
       "4                                     pT3                             pN0   \n",
       "\n",
       "  characteristics.tag.stage.mets characteristics.tag.primaryVSmets  \\\n",
       "0                            pM0                           Primary   \n",
       "1                            pMX                           Primary   \n",
       "2                            pM0                           Primary   \n",
       "3                            pM1                           Primary   \n",
       "4                            pM0                           Primary   \n",
       "\n",
       "  characteristics.tag.grade molecule tested   label  platform  \n",
       "0                         3       total RNA  biotin  GPL15048  \n",
       "1             Not Available       total RNA  biotin  GPL15048  \n",
       "2                         3       total RNA  biotin  GPL15048  \n",
       "3             Not Available       total RNA  biotin  GPL15048  \n",
       "4                         2       total RNA  biotin  GPL15048  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_excel('data/Lung3.metadata.xls')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bde83aae",
   "metadata": {},
   "source": [
    "## Data exploration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a52ef1",
   "metadata": {},
   "source": [
    "### Meta data exploration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7ffb2f9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(89, 16)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1e76d1d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 89 entries, 0 to 88\n",
      "Data columns (total 16 columns):\n",
      " #   Column                                          Non-Null Count  Dtype  \n",
      "---  ------                                          --------------  -----  \n",
      " 0   sample.name                                     89 non-null     object \n",
      " 1   title                                           89 non-null     object \n",
      " 2   CEL.file                                        89 non-null     object \n",
      " 3   source.location                                 89 non-null     object \n",
      " 4   organism                                        89 non-null     object \n",
      " 5   characteristics.tag.gender                      89 non-null     object \n",
      " 6   characteristics.tag.histology                   89 non-null     object \n",
      " 7   characteristics.tag.tumor.size.maximumdiameter  88 non-null     float64\n",
      " 8   characteristics.tag.stage.primary.tumor         89 non-null     object \n",
      " 9   characteristics.tag.stage.nodes                 89 non-null     object \n",
      " 10  characteristics.tag.stage.mets                  89 non-null     object \n",
      " 11  characteristics.tag.primaryVSmets               89 non-null     object \n",
      " 12  characteristics.tag.grade                       89 non-null     object \n",
      " 13  molecule tested                                 89 non-null     object \n",
      " 14  label                                           89 non-null     object \n",
      " 15  platform                                        89 non-null     object \n",
      "dtypes: float64(1), object(15)\n",
      "memory usage: 11.2+ KB\n"
     ]
    }
   ],
   "source": [
    "meta.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70066df4",
   "metadata": {},
   "source": [
    "Clearly, most of the meta-data is labels/catergorical data. Features such as organism are redundant considering it was done only to humans and other features also only have a single value/label. For this reason, 'organism', 'characteristics.tag.primaryVSmets', 'molecule tested', 'label' and 'platform' will be dropped. The first three rows also are only used for identification and can be dropped as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3681ccca",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_filtered = meta.drop(['sample.name',\n",
    "                           'title',\n",
    "                           'CEL.file',\n",
    "                           'organism',\n",
    "                          'characteristics.tag.primaryVSmets',\n",
    "                          'molecule tested',\n",
    "                          'label',\n",
    "                          'platform'], axis=1)\n",
    "\n",
    "meta_filtered.replace('Not Available', np.nan, inplace=True)\n",
    "meta_filtered['characteristics.tag.grade'] = meta_filtered['characteristics.tag.grade'].astype('float')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6b34a910",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "source.location                                    object\n",
       "characteristics.tag.gender                         object\n",
       "characteristics.tag.histology                      object\n",
       "characteristics.tag.tumor.size.maximumdiameter    float64\n",
       "characteristics.tag.stage.primary.tumor            object\n",
       "characteristics.tag.stage.nodes                    object\n",
       "characteristics.tag.stage.mets                     object\n",
       "characteristics.tag.grade                         float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_filtered.dtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2152b666",
   "metadata": {},
   "source": [
    "The meta-data currently is categorical, a statistical test cannot work with this, for this reason the labels will be mapped to values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a9a5bfd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source.location</th>\n",
       "      <th>characteristics.tag.gender</th>\n",
       "      <th>characteristics.tag.histology</th>\n",
       "      <th>characteristics.tag.tumor.size.maximumdiameter</th>\n",
       "      <th>characteristics.tag.stage.primary.tumor</th>\n",
       "      <th>characteristics.tag.stage.nodes</th>\n",
       "      <th>characteristics.tag.stage.mets</th>\n",
       "      <th>characteristics.tag.grade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.8</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>8.5</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>89 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    source.location  characteristics.tag.gender  \\\n",
       "0               0.0                         0.0   \n",
       "1               0.0                         0.0   \n",
       "2               0.0                         0.0   \n",
       "3               0.0                         0.0   \n",
       "4               0.0                         1.0   \n",
       "..              ...                         ...   \n",
       "84              3.0                         0.0   \n",
       "85              3.0                         0.0   \n",
       "86              4.0                         0.0   \n",
       "87              3.0                         0.0   \n",
       "88              2.0                         1.0   \n",
       "\n",
       "    characteristics.tag.histology  \\\n",
       "0                             0.0   \n",
       "1                             1.0   \n",
       "2                             2.0   \n",
       "3                             3.0   \n",
       "4                             0.0   \n",
       "..                            ...   \n",
       "84                            8.0   \n",
       "85                            0.0   \n",
       "86                           15.0   \n",
       "87                           16.0   \n",
       "88                            0.0   \n",
       "\n",
       "    characteristics.tag.tumor.size.maximumdiameter  \\\n",
       "0                                              4.0   \n",
       "1                                              1.3   \n",
       "2                                             11.0   \n",
       "3                                              NaN   \n",
       "4                                              7.8   \n",
       "..                                             ...   \n",
       "84                                             2.2   \n",
       "85                                             5.0   \n",
       "86                                             3.5   \n",
       "87                                             8.5   \n",
       "88                                             3.5   \n",
       "\n",
       "    characteristics.tag.stage.primary.tumor  characteristics.tag.stage.nodes  \\\n",
       "0                                       0.0                              0.0   \n",
       "1                                       1.0                              1.0   \n",
       "2                                       2.0                              0.0   \n",
       "3                                       3.0                              2.0   \n",
       "4                                       2.0                              0.0   \n",
       "..                                      ...                              ...   \n",
       "84                                     10.0                              0.0   \n",
       "85                                      2.0                              5.0   \n",
       "86                                      0.0                              0.0   \n",
       "87                                      2.0                              0.0   \n",
       "88                                      0.0                              4.0   \n",
       "\n",
       "    characteristics.tag.stage.mets  characteristics.tag.grade  \n",
       "0                              0.0                        3.0  \n",
       "1                              1.0                        NaN  \n",
       "2                              0.0                        3.0  \n",
       "3                              2.0                        NaN  \n",
       "4                              0.0                        2.0  \n",
       "..                             ...                        ...  \n",
       "84                             0.0                        NaN  \n",
       "85                             0.0                        3.0  \n",
       "86                             0.0                        2.0  \n",
       "87                             0.0                        NaN  \n",
       "88                             0.0                        1.0  \n",
       "\n",
       "[89 rows x 8 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mapping_dict(df):\n",
    "    \"\"\"\n",
    "    Creates a dict for remapping non-numeric columns.\n",
    "    \"\"\"\n",
    "    \n",
    "    col_dict = {}\n",
    "    for col in df.columns:\n",
    "        if df[col].dtypes != 'object':\n",
    "            pass\n",
    "        else:\n",
    "            u_vals = df[col].unique()\n",
    "            count = -1.0\n",
    "            unique_dict = {}\n",
    "            for val in u_vals:\n",
    "                count += 1.0\n",
    "                unique_dict[val] = count\n",
    "\n",
    "            col_dict[col] = unique_dict\n",
    "\n",
    "    return col_dict\n",
    "\n",
    "def df_remapper(df, map_dict):\n",
    "    \"\"\"\n",
    "    Uses a dictionairy to remap all non-numeric columns\n",
    "    \"\"\"\n",
    "    for col in df.columns:\n",
    "        if df[col].dtypes != 'object':\n",
    "            pass\n",
    "        else:\n",
    "            df[col] = df[col].map(map_dict[col])\n",
    "    return df\n",
    "\n",
    "\n",
    "mapdict = mapping_dict(meta_filtered)\n",
    "df_mapped = df_remapper(meta_filtered, mapdict)\n",
    "df_mapped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2fcda37",
   "metadata": {},
   "source": [
    "### Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "07c7b7a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(89, 60607)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cbc93145",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 89 entries, GSM1416528 to GSM1416616\n",
      "Columns: 60607 entries, AFFX-BioB-3_at to merck2-Z75331_a_at\n",
      "dtypes: float64(60607)\n",
      "memory usage: 41.2+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2a0ea72",
   "metadata": {},
   "source": [
    "There seem to be 89 rows and 60607 columns. This definitly is a very large number of dimensions and needs to be scaled down.\n",
    "\n",
    "A correlation matrix will be made to investigate correlation within the dataset. Highly correlated features can be removed as these only bring redundant information. This is preferably removed to create a simpler model with less features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8676a7a4",
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 27.4 GiB for an array with shape (60607, 60607) and data type float64",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_34509/3198735960.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m corr_matrix = df.corr(\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mmethod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'pearson'\u001b[0m\u001b[0;34m,\u001b[0m  \u001b[0;31m# Correlation method\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mmin_periods\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m      \u001b[0;31m# Min number of observations required\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m )\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mcorr\u001b[0;34m(self, method, min_periods)\u001b[0m\n\u001b[1;32m   9377\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   9378\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmethod\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"pearson\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 9379\u001b[0;31m             \u001b[0mcorrel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlibalgos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnancorr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mminp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmin_periods\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   9380\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mmethod\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"spearman\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   9381\u001b[0m             \u001b[0mcorrel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlibalgos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnancorr_spearman\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mminp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmin_periods\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.8/site-packages/pandas/_libs/algos.pyx\u001b[0m in \u001b[0;36mpandas._libs.algos.nancorr\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mMemoryError\u001b[0m: Unable to allocate 27.4 GiB for an array with shape (60607, 60607) and data type float64"
     ]
    }
   ],
   "source": [
    "# corr_matrix = df.corr(\n",
    "#     method = 'pearson',  # Correlation method\n",
    "#     min_periods = 1      # Min number of observations required\n",
    "# )\n",
    "\n",
    "# # set diagonal to 0.0 to remove self-correlation\n",
    "# np.fill_diagonal(corr_matrix.values, 0.0)\n",
    "\n",
    "# corr_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72ab0e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the most highlt correlated columns\n",
    "corr_matrix.abs().idxmax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd13676c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how hight the correlation is.\n",
    "corr_matrix.abs().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc1447e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# .skew 0: no skew, + right skew, - left skew, look for above .75 \n",
    "skew_columns = (df_unlabel.skew().sort_values(ascending=False))\n",
    "print('-- skew --')\n",
    "print(skew_columns)\n",
    "print('\\n')\n",
    "\n",
    "skew_columns_75 = skew_columns.loc[skew_columns > 0.75]\n",
    "print('-- skew > 0.75 --')\n",
    "print(skew_columns_75)\n",
    "\n",
    "print('-- max skew --')\n",
    "print(skew_columns.max())\n",
    "print('-- min skew --')\n",
    "print(skew_columns.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45316d13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot a histogram to view the data distribution\n",
    "plt.subplots(figsize =(12, 12))\n",
    "for col in df_unlabel.columns:\n",
    "    plt.hist(df_unlabel[col], bins='auto')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a95283d9",
   "metadata": {},
   "source": [
    "According to the above information, the dataset initially seemed to have a slight skew (skew in this dataset varies from 0.39 and 0.58) and after verifying it with a histogram, it clearly seems to be skewed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a910c6fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in df_unlabel.columns:\n",
    "    stats.probplot(df_unlabel[col], dist=\"norm\", plot=pylab)\n",
    "pylab.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee620b3b",
   "metadata": {},
   "source": [
    "## Data pre-processing\n",
    "\n",
    "To resolve the data skewness a number of techniques can be employed, the aim is to get a number as close as possible to zero, such there is minimal skewing in the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3459e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# log transform\n",
    "df_log = np.log(df_unlabel)\n",
    "print('Log transform skew: min: {:.3f}, max: {:.3f}'.format(df_log.skew().min(), df_log.skew().max()))\n",
    "df_log.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4e1f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# square root transform\n",
    "df_sqrt = np.sqrt(df_unlabel)\n",
    "print('Square root transform skew: min: {:.3f}, max: {:.3f}'.format(df_sqrt.skew().min(), df_sqrt.skew().max()))\n",
    "df_sqrt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fe221b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot a histogram to view the data distribution\n",
    "plt.subplots(figsize =(12, 12))\n",
    "for col in df_sqrt.columns:\n",
    "    plt.hist(df_sqrt[col], bins='auto')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f528ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in df_sqrt.columns:\n",
    "    stats.probplot(df_sqrt[col], dist=\"norm\", plot=pylab)\n",
    "pylab.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6419fc7b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
